File: analysis.py
import time
import random
from algorithms import dijkstra, bidirectional_dijkstra  # type: ignore

def compare_algorithms(graph, num_tests=10, seed=42):
    random.seed(seed)
    nodes = list(graph.adj.keys())
    
    if not nodes:
        print("Graph is empty!")
        return

    print(f"{'Source':<10} {'Target':<10} {'Dijkstra Dist':<15} {'Bi-Dir Dist':<15} {'Dijkstra Time(s)':<18} {'Bi-Dir Time(s)':<18}")
    print("-" * 90)
    
    total_dijkstra_time = 0
    total_bidir_time = 0
    
    for _ in range(num_tests):
        start = random.choice(nodes)
        end = random.choice(nodes)
        
        # Test Dijkstra
        t0 = time.time()
        d_dist, d_pred = dijkstra(graph, start, end)
        t_d = time.time() - t0
        dist_val = d_dist.get(end, float('inf'))
        total_dijkstra_time += t_d
        
        # Test Bidirectional
        t0 = time.time()
        b_dist, b_path = bidirectional_dijkstra(graph, start, end)
        t_b = time.time() - t0
        total_bidir_time += t_b
        
        print(f"{start:<10} {end:<10} {dist_val:<15} {b_dist:<15} {t_d:<18.4f} {t_b:<18.4f}")
        
    print("-" * 90)
    print(f"Average Dijkstra time:     {total_dijkstra_time/num_tests:.4f} s")
    print(f"Average Bidirectional time:{total_bidir_time/num_tests:.4f} s")

File: src/analysis.cpp
#include "analysis.h"
#include "algorithms.h"
#include <iostream>
#include <iomanip>
#include <random>
#include <chrono>
#include <vector>

void compare_algorithms(const Graph& graph, int num_tests, int seed) {
    if (graph.adj.empty()) {
        std::cout << "Graph is empty!\n";
        return;
    }
    
    std::mt19937 gen(seed);
    std::vector<int> nodes;
    for (const auto& pair : graph.adj) {
        nodes.push_back(pair.first);
    }
    
    std::uniform_int_distribution<> distrib(0, nodes.size() - 1);
    
    std::cout << std::left << std::setw(10) << "Source" 
              << std::setw(10) << "Target" 
              << std::setw(15) << "Dijkstra Dist" 
              << std::setw(15) << "Bi-Dir Dist" 
              << std::setw(18) << "Dijkstra Time(s)" 
              << std::setw(18) << "Bi-Dir Time(s)\n";
              
    std::cout << std::string(90, '-') << "\n";
    
    double total_dijkstra_time = 0.0;
    double total_bidir_time = 0.0;
    
    for (int i = 0; i < num_tests; ++i) {
        int start = nodes[distrib(gen)];
        int end = nodes[distrib(gen)];
        
        auto d_res = dijkstra(graph, start, end, true);
        float dist_val = d_res.distance;
        total_dijkstra_time += d_res.execution_time_ms / 1000.0;
        
        auto b_res = bidirectional_dijkstra(graph, start, end);
        total_bidir_time += b_res.execution_time_ms / 1000.0;
        
        std::cout << std::left << std::setw(10) << start 
                  << std::setw(10) << end 
                  << std::setw(15) << dist_val 
                  << std::setw(15) << b_res.distance
                  << std::setw(18) << std::fixed << std::setprecision(4) << (d_res.execution_time_ms / 1000.0) 
                  << std::setw(18) << std::fixed << std::setprecision(4) << (b_res.execution_time_ms / 1000.0) << "\n";
    }
    
    std::cout << std::string(90, '-') << "\n";
    std::cout << "Average Dijkstra time:     " << std::fixed << std::setprecision(4) << (total_dijkstra_time / num_tests) << " s\n";
    std::cout << "Average Bidirectional time:" << std::fixed << std::setprecision(4) << (total_bidir_time / num_tests) << " s\n";
}
